src/pks/flow/constitutive_relations/sources/pipe_drain_evaluator.cc

```diff
@@ -69,32 +177,65 @@ void PipeDrainEvaluator::Evaluate_(const State& S,
 
   int ncells = res.MyLength();
 
-  if(!pipe_domain_name_.empty()){
+  // generate cell maps
+  if (pipe_map_created == false){
+    CreateCellMap(S);
+    pipe_map_created = true;
+  }
+  if (sw_map_created == false) {
+    CreateCellMap(S);
+    sw_map_created = true;
+  }
+
+  double drain_length;
 
+  if(!pipe_domain_name_.empty()){
      const Epetra_MultiVector& pressHead = *S.GetPtr<CompositeVector>(pressure_head_key_, tag)
          ->ViewComponent("cell",false);
+     int c_pipe, c_sw;
 
      for (int c=0; c!=ncells; ++c) {
-        if (pressHead[0][c] < drain_length_) {
-           res[0][c] = - mnhMask[0][c] *  2.0 / 3.0 * energ_loss_coeff_weir_ * mnhPerimeter * sqrtTwoG * pow(srfcDepth[0][c],3.0/2.0);
+        // use cell map
+        if (pipe_flag == true) {
+          c_pipe = c;
+          c_sw = pipe_map[c];
+        } else if (sw_flag == true) {
+          c_sw = c;
+          c_pipe = sw_map[c];
+        }
+      
+        // calculate pipe drain length using bathymetry
+        drain_length = srfcB[0][c_sw] - pipeB[0][c_pipe];
+
+        // throw error if bathymetrys are not physical
+        if (drain_length < 0.0) {
+          std::cout<<"Pipe drain length negative; bathymetrys not physical"<<std::endl;
+          AMANZI_ASSERT(0);
+        }
+      
+        if (pressHead[0][c_pipe] < drain_length) {
+           res[0][c] = - mnhMask[0][c] *  2.0 / 3.0 * energ_loss_coeff_weir_ * mnhPerimeter * sqrtTwoG * pow(srfcDepth[0][c_sw],3.0/2.0);
         } 
-        else if (drain_length_ < pressHead[0][c] && pressHead[0][c] < (drain_length_ + srfcDepth[0][c]) ){
+        else if (drain_length < pressHead[0][c_pipe] && pressHead[0][c_pipe] < (drain_length + srfcDepth[0][c_sw]) ){
            res[0][c] = - mnhMask[0][c] * energ_loss_coeff_subweir_ * mnhArea * sqrtTwoG 
-                       * sqrt(srfcDepth[0][c] + drain_length_ - pressHead[0][c]);   
+                       * sqrt(srfcDepth[0][c_sw] + drain_length - pressHead[0][c_pipe]);   
         } 
-        else if (pressHead[0][c] > (drain_length_ + srfcDepth[0][c])) {
-           res[0][c] = mnhMask[0][c] * energ_loss_coeff_orifice_ * mnhArea * sqrtTwoG * sqrt(pressHead[0][c] - drain_length_ - srfcDepth[0][c]);
+        else if (pressHead[0][c_pipe] > (drain_length + srfcDepth[0][c_sw])) {
+           res[0][c] = mnhMask[0][c] * energ_loss_coeff_orifice_ * mnhArea * sqrtTwoG * sqrt(pressHead[0][c_pipe] - drain_length - srfcDepth[0][c_sw]);
         }
+     res[0][c] *= sink_source_coeff_;
      }
   }
   else {
      for (int c=0; c!=ncells; ++c) {
            res[0][c] = - mnhMask[0][c] *  2.0 / 3.0 * energ_loss_coeff_weir_ * mnhPerimeter * sqrtTwoG * pow(srfcDepth[0][c],3.0/2.0);
+           res[0][c] *= sink_source_coeff_;
      }
   }   
 
 }
 
+// Needs to be possibly UPDATED
```

Comment:
Please update the `EvaluatePartialDerivative_` as well.

Reply:
Done